source/slang/slang-ir.cpp

```diff
@@ -4287,11 +4289,117 @@ namespace Slang
         }
     }
 
+
+    void IRModule::trySearchForAndFillAllRayVariables(){
+        //instead of forcing order of code, a single pass if needed can be used
+        //likley forcing order is a faster and better idea
+
+        if(!checkedForRayVariables)
+        {
+            checkedForRayVariables = true;
+            for(auto i : getGlobalInsts())
+            {
+                if(auto rayDec = i->findDecoration<IRVulkanRayPayloadDecoration>())
+                {
+                    storeLocationToRayPayloadVariable(getIntVal(rayDec->getOperand(0)), i);
+                }
+                else if(auto rayDec = i->findDecoration<IRVulkanRayPayloadInDecoration>())
+                {
+                    storeLocationToRayPayloadVariable(getIntVal(rayDec->getOperand(0)), i);
+                }
+                else if(auto rayDec = i->findDecoration<IRVulkanHitObjectAttributesDecoration>())
+                {
+                    storeLocationToRayAttributeVariable(getIntVal(rayDec->getOperand(0)), i);
+                }
+                else if(auto rayDec = i->findDecoration<IRVulkanCallablePayloadDecoration>())
+                {
+                    storeLocationToRayCallableVariable(getIntVal(rayDec->getOperand(0)), i);
+                }
+                else if(auto rayDec = i->findDecoration<IRVulkanCallablePayloadInDecoration>())
+                {
+                    storeLocationToRayCallableVariable(getIntVal(rayDec->getOperand(0)), i);
+                }
+            }
+
+        }
+    }
+
+    IRInst* IRModule::getRayVariableFromLocation(IRInst* payloadVariable, Slang::IROp op, DiagnosticSink* sink)
+    {
+        IRBuilder builder(payloadVariable);
+        IRInst** varLayoutPointsTo;
+        int intLitValue = -1;
+        IRIntLit* intLit = as<IRIntLit>(payloadVariable);
+        if(intLit){
+            intLitValue = intLit->getValue();
+            if (kIROp_SPIRVAsmOperandRayPayloadFromLocation == op)
+            {
+                varLayoutPointsTo = this->getRayPayloadVariableFromLocation(intLitValue);
+            }
+            else if (kIROp_SPIRVAsmOperandRayAttributeFromLocation == op)
+            {
+                varLayoutPointsTo = this->getRayAttributeVariableFromLocation(intLitValue);
+            }
+            else if (kIROp_SPIRVAsmOperandRayCallableFromLocation == op)
+            {
+                varLayoutPointsTo = this->getRayCallableVariableFromLocation(intLitValue);
+            }
+            else
+            {
+                // should be impossible to hit
+                assert(false);
+            }
```

Comment:
I think this can be better with
```
else
{
    SLANG_ASSERT(kIROp_SPIRVAsmOperandRayCallableFromLocation == op);
    varLayoutPointsTo = this->getRayCallableVariableFromLocation(intLitValue);
}